## Exercises: Group and aggregate

In [1]:
import pandas as pd
import seaborn as sns
titanic = sns.load_dataset("titanic")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**Use group by and aggregation on the "titanic" dataset to answer the following questions:**

**Q1:** How many males and how many females were onboard titanic?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>males:</b> 577<br>
&nbsp;&nbsp;&nbsp;<b>females:</b> 314<br>
</details>

In [2]:
gender = titanic.groupby("sex")

# len(gender)

# titanic.loc[gender.groups["female"]]

# gender.get_group("female")

# gender.count()

gender["sex"].count()


sex
female    314
male      577
Name: sex, dtype: int64

**Q2:** How many males and how many females survived?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>males:</b> 109<br>
&nbsp;&nbsp;&nbsp;<b>females:</b> 233<br>
</details>

In [3]:
gender["survived"].count()

sex
female    314
male      577
Name: survived, dtype: int64

**Q3:** How many people traveled in each class, and what was the total amount spent on tickets in each class?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>First:</b> 216 people (£18177)<br>
&nbsp;&nbsp;&nbsp;<b>Second:</b> 184 people (£3802)<br>
&nbsp;&nbsp;&nbsp;<b>Third:</b> 491 people (£6715)<br>
</details>

In [4]:
class_tickets = titanic.groupby("class", observed=True)

display(class_tickets["sex"].count())

display(round(class_tickets["fare"].sum()))

class
First     216
Second    184
Third     491
Name: sex, dtype: int64

class
First     18177.0
Second     3802.0
Third      6715.0
Name: fare, dtype: float64

**Q4:** Where did most people embark, and what town had the highest ratio of survivors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;Most people (644) embarked in Southampton.<br>
&nbsp;&nbsp;&nbsp;However,Cherbourg had the highest ratio of survivors (55.4%).
</details>

In [5]:
embarked = titanic.groupby("embark_town")["sex"].count()

survivors = titanic.groupby("embark_town")["survived"].sum()

survivors_ratio = survivors/embarked * 100

display(embarked)
display(round(survivors_ratio, 1))

embark_town
Cherbourg      168
Queenstown      77
Southampton    644
Name: sex, dtype: int64

embark_town
Cherbourg      55.4
Queenstown     39.0
Southampton    33.7
dtype: float64

**Q5:** How many family members did survivors travel with on average? And for non-survivors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Survivors:</b> 0.94 family members<br>
&nbsp;&nbsp;&nbsp;<b>Non-Survivors:</b> 0.88 family members<br><br>
&nbsp;&nbsp;&nbsp;<i>* Family members counted as: sibsp + parch</i>
</details>

In [6]:
display(titanic.head(1))

survived_grouped = titanic.groupby("survived")

total_relatives = survived_grouped["sibsp"].sum() + survived_grouped["parch"].sum()         

count_survived = survived_grouped["sex"].count()      # Får tillbaka serie


round(total_relatives/count_survived ,2)    # går att göra operationer med serier 



survived  pclass   sex   age  sibsp  parch  fare embarked  class  who  \
0         0       3  male  22.0      1      0  7.25        S  Third  man   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False

survived
0    0.88
1    0.94
dtype: float64

**Use group by and aggregation on the "titanic" dataset to solve the following tasks:**

**T1:** Extract the following data from the dataset:
<!DOCTYPE html>
<html>
<body>

<table style="font-family: Arial, sans-serif; border-collapse: collapse; width: 50%; font-size: 12px;">
  <tr>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;"></th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">who</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_passengers</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_survivors</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">ratio_survivors</th>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">woman</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">271</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">205</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.756458</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">1</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">man</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">537</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">88</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.163873</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">2</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">child</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">83</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">49</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0.590361</td>
  </tr>
</table>

</body>
</html>
<br>
<details>
<summary>Solution</summary>
<pre><code>
titanic.groupby("who").aggregate(
    number_of_passengers=pd.NamedAgg(column="survived", aggfunc="count"),
    number_of_survivors=pd.NamedAgg(column="survived", aggfunc="sum"),
    ratio_survivors=pd.NamedAgg(column="survived", aggfunc="mean"),
).sort_index(ascending=False).reset_index()
</code></pre>
</details>

In [7]:
titanic.groupby("who").agg(
    number_of_passengers = pd.NamedAgg(column = "survived", aggfunc = "count"),      
    number_of_survivors = pd.NamedAgg(column = "survived", aggfunc = "sum"),
    ratio_survivors = pd.NamedAgg(column = "survived", aggfunc = "mean")         # istället för "max" eller "min", hade man kunnat skicka in en funktion se nästa ruta
).head(3)

number_of_passengers  number_of_survivors  ratio_survivors
who                                                              
child                    83                   49         0.590361
man                     537                   88         0.163873
woman                   271                  205         0.756458

**T2:** Create a custom function *percent()* that can be used to aggregate survivor and get a percentage string:

<!DOCTYPE html>
<html>
<body>

<table style="font-family: Arial, sans-serif; border-collapse: collapse; width: 50%; font-size: 12px;">
  <tr>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;"></th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">who</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_passengers</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">number_of_survivors</th>
    <th style="border: none; text-align: right; padding: 8px; background-color: #f2f2f2; color: black;">percentage_survivors</th>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">0</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">woman</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">271</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">205</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">75.6 %</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">1</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">man</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">537</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">88</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">16.4 %</td>
  </tr>
  <tr>
    <td style="border: none; text-align: right; padding: 8px; color: black;">2</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">child</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">83</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">49</td>
    <td style="border: none; text-align: right; padding: 8px; color: black;">59.0 %</td>
  </tr>
</table>
</body>
</html>

<br>
<details>
<summary>Solution</summary>
<pre><code>
import numpy as np
def percent(col):
    return f"{np.mean(col) * 100:.1f} %"
</code></pre>
</details>


In [35]:

def percent(series):
    """
    Custom function to aggregate survivors and return a percentage string.
    
    Parameters:
        - series: A pandas Series containing binary values (0 for not survived, 1 for survived).
        
    Returns:
        - A percentage string.
    """
    total_count = len(series)
    survived_count = series.sum()
    survival_rate = (survived_count / total_count) * 100
    
    return f"{survival_rate:.2f}%"


titanic.groupby("who").agg(
    number_of_passengers = pd.NamedAgg(column = "survived", aggfunc = "count"),      
    number_of_survivors = pd.NamedAgg(column = "survived", aggfunc = "sum"),
    ratio_survivors = pd.NamedAgg(column = "survived", aggfunc = percent)         # istället för "max" eller "min", hade man kunnat skicka in en funktion se nästa ruta
).head(3)



number_of_passengers  number_of_survivors ratio_survivors
who                                                             
child                    83                   49          59.04%
man                     537                   88          16.39%
woman                   271                  205          75.65%